### PR TITLE
[wip] check response body when veting masquerade

### PR DESCRIPTION
**Pull request by joesis**

### CODE REVIEW

File: direct.go [hunk 1]
```diff
@@ -223,11 +230,11 @@ func (d *direct) RoundTrip(req *http.Request) (*http.Response, error) {
 			masqueradeGood(false)
 			continue
 		}
-		resp.Body.Close()
 		if resp.StatusCode != http.StatusForbidden {
 			masqueradeGood(true)
 			return resp, nil
 		}
+		resp.Body.Close()
```
Comment (joesis) [inline direct.go]: This is the real fix: don't close body when there's no error.

File: direct_test.go [hunk 2]
```diff
@@ -29,8 +29,11 @@ func doTestDomainFronting(t *testing.T, cacheFile string) {
 	client := &http.Client{
 		Transport: NewDirect(30 * time.Second),
 	}
-	assert.True(t, doPostCheck(client))
-	log.Debugf("DIRECT DOMAIN FRONTING TEST SUCCEEDED")
+	assert.True(t, doCheck(client, http.MethodPost, testURL))
+	// client = &http.Client{
+	// 	Transport: NewDirect(30 * time.Second),
+	// }
+	// assert.True(t, doCheck(client, http.MethodGet, "http://d33pfmbpauhmvd.cloudfront.net/proxies.yaml.gz"))
```
Comment (joesis) [inline direct_test.go]: I still don't know why, but simply use this as `testURL` (without any other changes in this PR) fails the test.

Comment (oxtoacart) [inline direct_test.go]: Probably because you can't POST to that URL.

Comment (oxtoacart) [inline direct_test.go]: Actually, that shouldn't matter.

Comment (oxtoacart) [inline direct_test.go]: The POST is only used for checking masquerades.

File: direct.go [hunk 3]
```diff
@@ -167,27 +168,30 @@ func postCheck(conn net.Conn) bool {
...
 }
 
-func doCheck(client *http.Client, method string, u string) bool {
-	req, _ := http.NewRequest(method, u, strings.NewReader("a"))
-	req.Header.Set("Content-Type", "application/json")
+func doCheck(client *http.Client, method string, expectedStatus int, u string) bool {
+	isPost := method == http.MethodPost
+	var requestBody io.Reader
+	if isPost {
+		requestBody = strings.NewReader("a")
```
Comment (oxtoacart) [inline direct.go]: Only include body when doing a POST

File: direct.go [hunk 4]
```diff
@@ -167,27 +168,30 @@ func postCheck(conn net.Conn) bool {
...
+	}
+	req, _ := http.NewRequest(method, u, requestBody)
+	if isPost {
+		req.Header.Set("Content-Type", "application/json")
```
Comment (oxtoacart) [inline direct.go]: Only include content-type header when doing a POST

File: direct.go [hunk 5]
```diff
@@ -167,27 +168,30 @@ func postCheck(conn net.Conn) bool {
...
+	}
 	resp, err := client.Do(req)
 	if err != nil {
-		log.Tracef("Unsuccessful vetting with POST request, discarding masquerade: %v", err)
+		log.Tracef("Unsuccessful vetting with %v request, discarding masquerade: %v", method, err)
```
Comment (oxtoacart) [inline direct.go]: Show actual method.

File: direct.go [hunk 6]
```diff
@@ -167,27 +168,30 @@ func postCheck(conn net.Conn) bool {
...
 		return false
 	}
-	defer resp.Body.Close()
-	length, _ := strconv.Atoi(resp.Header.Get("Content-Length"))
-	if length > 0 {
-		read, _ := ioutil.ReadAll(resp.Body)
-		if len(read) != length {
-			return false
-		}
+	if resp.Body != nil {
```
Comment (oxtoacart) [inline direct.go]: Responses won't always have a Content-Length (for example if they're using chunked transfer encoding). Just check for non-nil body and discard here.

File: direct.go [hunk 7]
```diff
@@ -167,27 +168,30 @@ func postCheck(conn net.Conn) bool {
...
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
 	}
-	if resp.StatusCode != http.StatusAccepted {
-		log.Tracef("Unexpected response status vetting masquerade: %v, %v", resp.StatusCode, resp.Status)
+	if resp.StatusCode != expectedStatus {
```
Comment (oxtoacart) [inline direct.go]: The expected status depends on the endpoint, so we pass it in.

**Comment by oxtoacart** [comment]
@joesis Thanks for fixing this! I made a couple of tweaks, merging now.

**Comment by joesis** [comment]
Thanks @oxtoacart !

**Pull request merged**